test/integration/integration_test.go

```diff
@@ -0,0 +1,294 @@
+	err = sdk.Create(appService)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = sdk.Create(clusterDNS)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	clusterDomain := getClusterDomain(clusterDNS)
+	serviceIP, servicePort, err := getClusterIP(dnsService)
+	if err != nil {
+		t.Fatal(err)
+	}
+	t.Logf("configured cluster domain: %s, clusterIP: %s:%s", clusterDomain, serviceIP, servicePort)
+
+	testPod, err := getDNSPod(dnsNamespace.Name)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	type dnsTest struct {
+		domain  string
+		dnsType string
+	}
+	tests := []dnsTest{
+		{
+			domain:  fmt.Sprintf("dns-version.%s.", clusterDomain),
+			dnsType: "txt",
+		},
+		{
+			domain:  fmt.Sprintf("kubernetes.default.svc.%s.", clusterDomain),
+			dnsType: "a",
+		},
+		{
+			domain:  fmt.Sprintf("%s.%s.svc.%s.", appService.Name, appService.Namespace, clusterDomain),
+			dnsType: "a",
+		},
+		{
+			domain:  fmt.Sprintf("_https._tcp.kubernetes.default.svc.%s.", clusterDomain),
+			dnsType: "srv",
+		},
+		{
+			domain:  fmt.Sprintf("_http._tcp.%s.%s.svc.%s.", appService.Name, appService.Namespace, clusterDomain),
+			dnsType: "srv",
+		},
+	}
+
+	for _, test := range tests {
+		digCmd := fmt.Sprintf("dig +noall +answer %s %s @%s -p %s", test.dnsType, test.domain, serviceIP, servicePort)
+		cmd := fmt.Sprintf("oc exec %s -n %s -- %s", testPod, dnsNamespace.Name, digCmd)
+		msg := fmt.Sprintf("test domain: %s, type: %s", test.domain, test.dnsType)
+
+		outStr, errStr := runShellCmd(cmd, msg)
+		if len(errStr) != 0 {
+			t.Fatalf("failed to %s, err: %s", msg, errStr)
+		} else if len(outStr) == 0 {
+			t.Fatalf("failed to %s, invalid answer", msg)
+		}
+	}
+}
+
+func getClusterDomain(dns *dnsv1alpha1.ClusterDNS) string {
+	clusterDomain := "cluster.local"
+	if dns.Spec.ClusterDomain != nil {
+		clusterDomain = *dns.Spec.ClusterDomain
+	}
+	return clusterDomain
+}
+
+func getClusterIP(dnsService *corev1.Service) (string, string, error) {
+	var serviceIP string
+	var servicePort string
+	var err error
+
+	err = wait.Poll(500*time.Millisecond, time.Minute, func() (bool, error) {
+		err = sdk.Get(dnsService)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return false, nil
+			}
+			return false, err
+		}
+		if len(dnsService.Spec.ClusterIP) > 0 {
+			serviceIP = dnsService.Spec.ClusterIP
+			servicePort = strconv.FormatUint(uint64(dnsService.Spec.Ports[0].Port), 10)
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil {
+		return "", "", fmt.Errorf("timed out waiting for DNS service: %v", err)
+	}
+	return serviceIP, servicePort, nil
+}
+
+func getDNSPod(ns string) (string, error) {
+	var podName string
+
+	var err error
+	kubeClient := kubeclient.GetKubeClient()
+
+	err = wait.Poll(500*time.Millisecond, time.Minute, func() (bool, error) {
+		// NOTE: kubeClient is used instead of sdk.List(ns, podList) as the
+		// latter is not reliably populating pod.Status.Phase field at this time.
+		podList, err := kubeClient.Core().Pods(ns).List(metav1.ListOptions{})
```

Comment:
To make sure this isn't flaky, probably want to add in the DaemonSet selector or something. There might be some upstream util for this